htmresearch/frameworks/nlp/imbu.py

```diff
@@ -0,0 +1,203 @@
+import argparse
+import os
+
+from enum import Enum
+
+from htmresearch.encoders import EncoderTypes
+from htmresearch.frameworks.nlp.classification_model import ClassificationModel
+from htmresearch.frameworks.nlp.classify_fingerprint import (
+  ClassificationModelFingerprint)
+from htmresearch.frameworks.nlp.classify_htm import ClassificationModelHTM
+from htmresearch.frameworks.nlp.classify_keywords import (
+  ClassificationModelKeywords)
+from htmresearch.frameworks.nlp.classify_windows import (
+  ClassificationModelWindows)
+from htmresearch.support.csv_helper import readCSV
+from htmresearch.support.text_preprocess import TextPreprocess
+
+
+
+class ImbuModelTypes(Enum):
+  CioWordFingerprint = ClassificationModelFingerprint
+  CioDocumentFingerprint = ClassificationModelFingerprint
+  CioWindows = ClassificationModelWindows
+  Keywords = ClassificationModelKeywords
+  HTMNetwork = ClassificationModelHTM
+
+
+
+class ModelSimilarityMetrics(Enum):
+  pctOverlapOfInput = "pctOverlapOfInput"
+
+
+
+class ImbuError(Exception):
+  pass
+
+
+
+class ImbuUnableToLoadModelError(ImbuError):
+  def __init__(self, exc):
+    self.exc = exc
+
+
+
+class ImbuModels(object):
+
+  defaultSimilarityMetric = ModelSimilarityMetrics.pctOverlapOfInput
+  defaultModelType = ImbuModelTypes.HTMNetwork
+
+  def __init__(self, cacheRoot, dataPath, loadPath, savePath,
+      modelSimilarityMetric):
+    self.cacheRoot = cacheRoot
+    self.modelSimilarityMetric = (
+      modelSimilarityMetric or self.defaultSimilarityMetric
+    )
+    self.loadPath = loadPath
+    self.savePath = savePath
+    self.dataPath = dataPath
+    self.dataDict = self._loadTrainingData()
+
+
+  def __repr__(self):
+    return ("Imbu<cacheRoot={cacheRoot}, dataPath={dataPath}, "
+            "loadPath={loadPath}, savePath={savePath}, "
+            "modelSimilarityMetric={modelSimilarityMetric}>"
+            .format(**self.__dict__))
+
+
+  def _defaultModelFactoryKwargs(self):
+    return dict(
+      numLabels=len(self.dataDict),
+      modelDir=self.savePath,
+      classifierMetric=self.modelSimilarityMetric.value)
+
+
+  def modelFactory(self, modelType, **kwargs):
```

Comment:
Please also see frameworks/nlp/model_factory.py. Perhaps we should combine the two?